lib/rucio/rse/protocols/protocol.py

```diff
@@ -221,6 +222,39 @@ def __xenon(scope, name, rse, rse_attrs, protocol_attrs):
 
         return '%s/%s/%s/%s' % (scope[0:7], scope[4:len(scope)], name.split('-')[0] + "-" + name.split('-')[1], name)
 
+    @staticmethod
+    def __lsst(scope, name, rse_id, rse_attrs, protocol_attrs):
+        """
+        LFN2PFN algorithm for LSST
+        
+        Get any of the file replica and extract its relative path
+    
+        :param scope: Scope of the LFN.
+        :param name: File name of the LFN.
+        :param rse_id: RSE for PFN (ignored)
+        :param rse_attrs: RSE attributes for PFN (ignored)
+        :param protocol_attrs: RSE protocol attributes for PFN (ignored)
+        :returns: Path for use in the PFN generation.
+        """
+        del rse_id
+        del rse_attrs
+        del protocol_attrs
+        try:
+            from rucio.api import replica
```

Comment:
something is not right here. this function is potentially called thousands of times, so it shouldn't call list_replicas() or something similar. what exactly are you trying to achieve?

Reply:
This code will be used for replication only. When replicating a file from a RSE to another, we want to make sure the file will keep the same relative path (this path will originally be specified when uploading on a non-deterministic RSE).
Example : the file **RSE-SOURCE/root_source/the/relative/path/to/the/file** will be replicated on RSE-DEST at **RSE-DEST/root_dest/the/relative/path/to/the/file**
So the algo on RSE-DEST needs to get a replica of the file to retrieve the relative path.
It seems input params on the LFN2PFN function are not sufficent to do this without calling replica.get_replicas().
If you think there is a simpler solution, please let me know

Reply:
hmmm, isn't CMS doing it like this? @bari12 @ericvaandering mabey there's an easier way

Reply:
I don't 100% follow what you're doing, but CMS has two forms of doing this, one is a simplification of the other:

Form 1: PFN at site = RSE_SPECIFIC_PREFIX + '/' + LFN

Form 2: Pass LFN through a series of REGEX substitution processes to arrive at PFN (useful for changing the prefix of the PFN based on the LFN). 

In both cases, we actually discard (but do some checking) on the scope.

Both the RSE specific prefix and the regular expressions all come from RSE attributes (so we don't ignore those).

Reply:
It's clear that we cannot do like CMS: we don't have enough information in the LFN itself to construct the PFN. If the algo can't call API methods like list_replicas(), we have a problem here.
Let me explain our use-case:
Say we have the DID **scope:lsst_a_262593_R43_S02_z.fits** that was uploaded in the first place on a RSE-SOURCE (non-deterministic) with the full path **datasets/desc/DC2/Run2.2i/sim/y1-wfd/00262593/lsst_a_262593_R43_S02_z.fits**. The PFN of the replica is then **RSE-SOURCE/root_source/datasets/desc/DC2/Run2.2i/sim/y1-wfd/00262593/lsst_a_262593_R43_S02_z.fits**
We want to replicate it on a RSE-DEST with the same relative path ie we need to build a PFN like **RSE-DEST/root_dest/datasets/desc/DC2/Run2.2i/sim/y1-wfd/00262593/lsst_a_262593_R43_S02_z.fits**

Reply:
This won't work like this and goes against the fundamental idea of the deterministic lfn2pfn algorithms. (We should probably put up a page to describe this better somewhere)
The deterministic lfn2pfn algorithms should be able to generate the pfn purely based on the scope and name of a file (+ information like hostname, prefixes etc.) - no other lookup must be necessary. So your `list_replicas` call here breaks this concept.

I think the usecase you are describing calls for a specific surl algorithm. https://github.com/rucio/rucio/blob/4ff378c848ccc88f3558500bf17004c79a02765e/lib/rucio/common/utils.py#L578
What the SURL algorithm does is to calculate the path for transfers to non-deterministic RSEs. Since it is non-deterministic, Rucio cannot use the lfn2pfn algorithm here, but because it is an async transfer, it also cannot wait for user input (Like you would do on a file registration on a non-deterministic RSE). Thus there are these surl algorithms which still can programatically calculate the pfn for non-deterministic RSEs, but in contrast to the lfn2pfn methods, they are allowed to use additional knowledge to do so.

Reply:
Maybe I missing something, but why not make the DID look like this:

scope=`lsst` name=`/DC2/Run2.2i/sim/y1-wfd/00262593/lsst_a_262593_R43_S02_z.fits`

or

scope=`/DC2/Run2.2i/sim` name=`/y1-wfd/00262593/lsst_a_262593_R43_S02_z.fits`

? The whole point of this mapping is that you don't have to look it up in a database but that you can compute it offline.

Reply:
> This won't work like this and goes against the fundamental idea of the deterministic lfn2pfn algorithms. (We should probably put up a page to describe this better somewhere) The deterministic lfn2pfn algorithms should be able to generate the pfn purely based on the scope and name of a file (+ information like hostname, prefixes etc.) - no other lookup must be necessary. So your `list_replicas` call here breaks this concept.
> 
OK I was not aware of this, I understand.

Reply:
> Maybe I missing something, but why not make the DID look like this:
> 
> scope=`lsst` name=`/DC2/Run2.2i/sim/y1-wfd/00262593/lsst_a_262593_R43_S02_z.fits`
> 
> or
> 
> scope=`/DC2/Run2.2i/sim` name=`/y1-wfd/00262593/lsst_a_262593_R43_S02_z.fits`
> 
> ? The whole point of this mapping is that you don't have to look it up in a database but that you can compute it offline.

Good point. This means we have to build the appropriate DID.scope and DID.name when we first upload the file, but I guess it is not big issue.
I then have a try with this solution and using the 'identity' algo.
Thanks for your help and sorry for the newbie questions ;)

Reply:
@mlassnig It seems the '/' is not allowed in the DID name (like in your proposal) by `rucio upload`:

```
$ rucio upload --rse IN2P3-CC-LSST-SOURCE --scope LSST_ESCAPE_SCHWARZ_test --name project.03 file
2021-10-05 17:07:29,960	INFO	Preparing upload for file .project
...
2021-10-05 17:07:32,448	INFO	Successfully uploaded file .project

$ rucio upload --rse IN2P3-CC-LSST-SOURCE --scope LSST_ESCAPE_SCHWARZ_test --name '/directory/project.03' file
2021-10-05 17:07:53,833	INFO	Preparing upload for file .project
2021-10-05 17:07:54,027	ERROR	An unknown exception occurred.
Details: no error information passed (http status code: 404)
```

any suggestions?

Reply:
The schema of your Rucio instance must allow it, it's not allowed in the generic (standard) schema, and it seems not in the ESCAPE one either.

Reply:
Digging a bit more in the Rucio API, I tried to register a file which already exists in the storage backend and give it a DID name with a '/' in it:

```
>>> from rucio.client.replicaclient import ReplicaClient
>>> rc = ReplicaClient()
>>> result = rc.add_replica(rse="IN2P3-CC-LSST-SOURCE",scope="LSST_ESCAPE_SCHWARZ_test",name="subdir/project",bytes=366,adler32="2b5b78fc")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.6/dist-packages/rucio/client/replicaclient.py", line 254, in add_replica
    return self.add_replicas(rse=rse, files=[dict])
  File "/usr/local/lib/python3.6/dist-packages/rucio/client/replicaclient.py", line 274, in add_replicas
    raise exc_cls(exc_msg)
rucio.common.exception.InvalidObject: Provided object does not match schema.
Details: Problem validating dids : 'subdir/project' does not match '^[A-Za-z0-9][A-Za-z0-9\\.\\-\\_]{1,250}$'

Failed validating 'pattern' in schema['items']['properties']['name']:
    {'description': 'Data Identifier name',
     'pattern': '^[A-Za-z0-9][A-Za-z0-9\\.\\-\\_]{1,250}$',
     'type': 'string'}

On instance[0]['name']:
    'subdir/project'
```

It seems the DID.name is limited to 250 chars and cannot contains chars other than alphanumeric and . - _

Reply:
> The schema of your Rucio instance must allow it, it's not allowed in the generic (standard) schema, and it seems not in the ESCAPE one either.

good to know this can be customized. just by curiosity, how easy is it to change this. Is this documented in https://rucio.readthedocs.io/en/latest ?